clippy_lints/src/question_mark.rs

```diff
@@ -39,135 +42,185 @@ declare_clippy_lint! {
-                        replacement = Some(format!("Some({}?)", receiver_str));
-                    }
-                } else if Self::moves_by_default(cx, subject)
-                    && !matches!(subject.kind, ExprKind::Call(..) | ExprKind::MethodCall(..))
-                {
-                    replacement = Some(format!("{}.as_ref()?;", receiver_str));
-                } else {
-                    replacement = Some(format!("{}?;", receiver_str));
-                }
+    /// Contains: let_pat_qpath (Xxx), let_pat_type, let_pat_sym (a), let_expr (b), if_then (c),
+    /// if_else (d)
+    IfLet(
+        &'hir QPath<'hir>,
+        Ty<'hir>,
+        Symbol,
+        &'hir Expr<'hir>,
+        &'hir Expr<'hir>,
+        Option<&'hir Expr<'hir>>,
+    ),
+}
 
-                if let Some(replacement_str) = replacement {
-                    span_lint_and_sugg(
-                        cx,
-                        QUESTION_MARK,
-                        expr.span,
-                        "this block may be rewritten with the `?` operator",
-                        "replace it with",
-                        replacement_str,
-                        applicability,
-                    );
+/// Checks if the given expression on the given context matches the following structure:
+///
+/// ```ignore
+/// if option.is_none() {
+///    return None;
+/// }
+/// ```
+///
+/// ```ignore
+/// if result.is_err() {
+///     return result;
+/// }
+/// ```
+///
+/// If it matches, it will suggest to use the question mark operator instead
+fn check_is_none_or_err_and_early_return(cx: &LateContext<'_>, expr: &Expr<'_>) {
+    if_chain! {
+        if let Some(higher::If { cond, then, r#else }) = higher::If::hir(expr);
+        if !is_else_clause(cx.tcx, expr);
+        if let ExprKind::MethodCall(segment, args, _) = &cond.kind;
+        if let Some(caller) = args.get(0);
+        let caller_ty = cx.typeck_results().expr_ty(caller);
+        let if_block = IfBlockType::IfIs(caller, caller_ty, segment.ident.name, then, r#else);
+        if is_early_return(sym::Option, cx, &if_block) || is_early_return(sym::Result, cx, &if_block);
+        then {
+            let mut applicability = Applicability::MachineApplicable;
+            let receiver_str = snippet_with_applicability(cx, caller.span, "..", &mut applicability);
+            let by_ref = !caller_ty.is_copy_modulo_regions(cx.tcx.at(caller.span), cx.param_env) &&
+                !matches!(caller.kind, ExprKind::Call(..) | ExprKind::MethodCall(..));
+            let mut suggestion = String::new();
+            if let Some(else_inner) = r#else {
+                if eq_expr_value(cx, caller, peel_blocks(else_inner)) {
+                    suggestion = format!("Some({}?)", receiver_str);
                 }
+            } else {
+                suggestion = format!("{}{}?;", receiver_str, if by_ref { ".as_ref()" } else { "" });
             }
```

Comment:
nit: Now that we're using let-chains, we can probably do:

```rust
if let Some(else_inner) = r#else && eq_expr_value(cx, caller, peel_blocks(else_inner) {
	suggestion = format!("Some({}?)", receiver_str);
} else {
	suggestion = format!("{}{}?;", receiver_str, if by_ref { ".as_ref()" } else { "" });
}
```

Reply:
finally~ I saw a pr which introducing it at the time, and I think I even posted a commit changing all the if-chains to let-chains, but I eventually revert back after that pr was closed due to concerns of formatting etc.

btw I'll get back to this pr once I got my hard drive fixed 🤦‍♂️

Reply:
> nit: Now that we're using let-chains, we can probably do:
> 
> ```rust
> if let Some(else_inner) = r#else && eq_expr_value(cx, caller, peel_blocks(else_inner) {
> 	suggestion = format!("Some({}?)", receiver_str);
> } else {
> 	suggestion = format!("{}{}?;", receiver_str, if by_ref { ".as_ref()" } else { "" });
> }
> ```

*hooray! finally got a new hard drive*

well, I did that but only realized it breaks one of the original tests, like when the return in `else` block is different than the caller:

```rust
if Some(1).is_none() {
    return None;
} else {
    return Some(0);
}
```

So I guess... I can't change too mach on this part?

Reply:
Oh wait, you're right. They're not exactly the same. :facepalm: